### PR TITLE
DOCS: Add Sphinx release notes, remove manual Changelog

### LINKED ISSUE #3095: MAINT: remove changelog, publish GH release notes
@thatlittleboy I have a proposal about our changelog - I think we can streamline the process a bit.

I originally added a changelog to the SHAP Community Fork, as we weren't making releases and needed a place to put changelog info. However, now that we're back in the main repo, we could remove the changelog and use the GitHub release notes to source the summarised changes (which is what has been done for most of the repo's history).

Some advantages:

1. **Commitless releases**
   - A new commit would not be needed to publish, which streamlines making a new release
   - We'd also need to source the version number from git, as in #3059.
2. **Auto-generated notes**
   - GitHub drafts the release notes messages from PR titles, which one can then manually tidy up and group into "Fix / Added / Changed / etc" subheadings
3. **Less admin on each PR** / simpler checklist

To make the release notes a bit more navigable, we could look at including the release notes in the docs, with [Sphinx Github Changelog](https://sphinx-github-changelog.readthedocs.io/en/latest/). If I get some time next week I'll give it a shot and see how it looks.

We could also make life easier for ourselves by ensuring the names of PRs are clear, if needed by **renaming the PRs before merging**. It's very helpful when the title includes with a verb like "Fix", "Add", "Remove", as that helps organise the messages into useful headings in the release notes. (It doesn't have to be perfect, as one can always reword things in the release notes themselves).

### PR DESCRIPTION
**Pull request by connortann**
## Summary

Closes #3095

- Adds release notes to the built documentation, automatically built from GitHub releases
- Removes the manual Changelog file (yay!)
- Simplifies Pull Request template 

## Preview

Docs can be previewed here: <https://shap--3104.org.readthedocs.build/en/3104/>

## Discussion

I feel quite positive about this PR, I think it a neat improvement to the docs and significantly reduces changelog faff :)

I believe this meets all the desired criteria in this comment: https://github.com/shap/shap/issues/3095#issuecomment-1637051807

> If not in CHANGELOG.md, then the changelog should be in the docs. ✔️

Done

> Ability to customize the changelog entry messages ✔️

The GitHub release notes are fully customisable. The auto-generated release notes from Github are great, they do 90% of the work: they link the PR and credit the author. One can edit them before publishing, and indeed we can edit them further after publishing if desired.

> Ability to exclude certain PRs from the changelog ✔️

 Yep, we can do that.

> The unreleased section ✔️

I think we can get rid of the previous manual list. As a more automated solution, one can compare `master` against the most recent release on GitHub e.g. <https://github.com/shap/shap/compare/v0.42.1...master>

I found a way to include this URL in the release notes. One challenge is that we need to know the latest git tag (e.g. `v0.42.1`) to build that URL. I used the GitHub API to get this dyamically, and the `rst_epilog` configuration to save the URL. 

> Proper categorization of each change in the CHANGELOG

Currently done manually when making a release, which only takes a minute or so. Automating with labels is a cool idea. Suggest implementing in a separate PR, as this already has a fair few changes. Side note: inspired by the latest default set of GitHub labels, I recently created labels in our repo for `duplicate`, `stale`, `good first issue` and also `javascript`.

### CODE REVIEW

File: docs/release_notes.rst [hunk 1]
```diff
@@ -0,0 +1,9 @@
+Release notes
+-------------
+
+To see the latest changes that are due on the next release, see |unreleasedchanges|_
+
+.. changelog::
+    :changelog-url: https://shap.readthedocs.io/en/latest/release_notes.html
```
Comment (thatlittleboy) [inline docs/release_notes.rst:7]: I'll like to also propose to add this release_notes page link to our `pyproject.toml`'s `[project.urls]` section as well.

I find being able to jump to changelogs from the PyPI page very convenient and useful (it saves at least 1 level of redirection).

File: docs/release_notes.rst [hunk 2]
```diff
@@ -0,0 +1,9 @@
+Release notes
+-------------
+
+To see the latest changes that are due on the next release, see |unreleasedchanges|_
```
Comment (thatlittleboy) [inline docs/release_notes.rst]: We'll need a period at the end of this sentence. :)

**Review by thatlittleboy — APPROVED** [review]
LGTM!

Let's see how it goes when we make our next release. I'm still reserved on whether this reduces or increases our workload overall. Especially when we have many PRs that shouldn't make it into the changelog (`dependabot` PRs, I'm looking at you 👀) and we'll have to remove them manually.

Let's explore the `labels` approach as a follow-up to this PR~ I've noticed many mature projects use this method and I think it's quite elegant.

fake-paws-dog-clapping

File: docs/release_notes.rst (hunk 1 above)
Comment (connortann) [reply docs/release_notes.rst:7]: Good idea, done

**Comment by connortann** [comment]
> Let's see how it goes when we make our next release. I'm still reserved on whether this reduces or increases our workload overall. Especially when we have many PRs that shouldn't make it into the changelog (`dependabot` PRs, I'm looking at you 👀) and we'll have to remove them manually.

I feel quite confident it will lower the workload overall, as it should be much faster to auto-generate the list and remove unwanted items than to write the list in the first place. In any case, let's see how it goes.
 
> Let's explore the `labels` approach as a follow-up to this PR~ I've noticed many mature projects use this method and I think it's quite elegant.

Sounds good :)

**Pull request merged**